python/ray/autoscaler/_private/aws/events.py

```diff
@@ -0,0 +1,88 @@
+        self.parameters = events_config.get("parameters", {})
+
+    def add_callback(self, event: CreateClusterEvent):
+        """Adds a callback handler based on the ARN that is supplied in `events.notification_uri`.
+        Currently, only SNS topics are supported.
+
+        Args:
+            event: The cluster event to invoke the callback handler for
+        """
+        if self.uri.startswith("arn:aws:sns"):
+            global_event_system.add_callback_handler(event,
+                                                     self._sns_callback,
+                                                     SnsHelper(self._get_region()),
+                                                     **self.parameters)
+        elif self.uri.startswith("arn:aws:lambda"):
+            global_event_system.add_callback_handler(event, self._lambda_callback)
+        elif self.uri.startswith("arn:aws:logs"):
+            global_event_system.add_callback_handler(event, self._cloudwatch_callback)
+        elif self.uri.startswith("arn:aws:apigateway"):
+            global_event_system.add_callback_handler(event, self._api_gateway_callback)
+
+    def _get_region(self):
+        return self.uri.split(":")[3]
+
+    def _sns_callback(self, sns_client: SnsHelper, event_data: Dict[str, Any], **kwargs):
+        """SNS callback for sending Ray cluster event data to an SNS topic.
+
+        Args:
+            sns_client: Amazon SNS client for publishing to an SNS topic
+            event_data: Ray cluster setup event data. This contains the event name, enum ID, and
+                may also contain additional metadata (i.e. the initialization or setup command used
+                during this setup step)
+            **kwargs: Keyword arguments that were injected into `_EventSystem.add_callback_handler`
+        """
+        try:
+            # create a copy of the event data to modify
+            event_dict = copy.deepcopy(event_data)
+            event: CreateClusterEvent = event_dict.pop("event_name")
+            node_context: NodeContext = event_dict.get("node_context", {})
+            sns_topic_arn, params = self.uri, kwargs
+            message = {
+                **params,
+                "setupEventMetadata": event_dict,
+                "stateSequence": event.value - 1,  # zero-index sequencing
+                "timestamp": round(time.time() * 1000)
+            }
+
+            if node_context:
+                message["rayNodeId"] = node_context["node_id"]
+                message["rayNodeType"] = "HEAD" if node_context["is_head_node"] else "WORKER"
+
+            sns_client.publish(sns_topic_arn, json.dumps(message))
+            logger.info("Published SNS event {} to {}".format(
+                event.name, sns_topic_arn))
+        except ClientError as exc:
+            cli_logger.abort(
+                "Failed to execute callback for create cluster events: {}", exc)
```

Comment:
Minor: cli_logger.abort(
                "{} Error caught when publishing {} create cluster events to SNS", exc.response["Error"], event.name)

Reply:
Updated!